plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,284 @@
+package t128_graphql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Jeffail/gabs"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	//make this configurable?
+	DefaultRequestTimeout = time.Second * 5
+)
+
+type T128GraphQL struct {
+	CollectorName string            `toml:"collector_name"`
+	BaseURL       string            `toml:"base_url"`
+	EntryPoint    string            `toml:"entry_point"`
+	Fields        map[string]string `toml:"extract_fields"`
+	Tags          map[string]string `toml:"extract_tags"`
+
+	Client *http.Client
+	Query  string
+}
+
+var sampleConfig = `
+## Collect data using graphQL
+[[inputs.t128_graphql]]
+## Required. The telegraf collector name
+# collector_name = "arp-state"
+
+## Required. GraphQL ports vary across 128T versions
+# base_url = "http://localhost:31517"
+
+## The starting point in the graphQL tree for all configured tags and fields
+# entry_point = "allRouters[name:RTR_WEST_COMBO]/nodes/nodes[name:combo-west]/nodes/arp/nodes"
+
+## The fields to collect with the desired name as the key (left) and graphQL 
+## key as the value (right)
+# [inputs.t128_graphql.extract_fields]
+#   state = "state"
+
+## The tags for filtering data with the desired name as the key (left) and 
+## graphQL key as the value (right)
+# [inputs.t128_graphql.extract_tags]
+#   networkInterface = "networkInterface"
+#   deviceInterface = "deviceInterface"
+#   vlan = "vlan"
+#   ipAddress = "ipAddress"
+#   destinationMac = "destinationMac"
+`
+
+func (*T128GraphQL) SampleConfig() string {
+	return sampleConfig
+}
+
+func (*T128GraphQL) Description() string {
+	return "Read data from a graphQL query"
+}
+
+func (plugin *T128GraphQL) Init() error {
+	if plugin.CollectorName == "" {
+		return fmt.Errorf("collector_name is a required configuration field")
+	}
+
+	if plugin.BaseURL == "" {
+		return fmt.Errorf("base_url is a required configuration field")
+	}
+
+	if plugin.BaseURL[len(plugin.BaseURL)-1:] != "/" {
+		plugin.BaseURL += "/"
+	}
+
+	transport := http.DefaultTransport
+
+	plugin.Client = &http.Client{Transport: transport, Timeout: DefaultRequestTimeout}
+	plugin.Query = plugin.buildQuery()
+
+	return nil
+}
+
+func (plugin *T128GraphQL) Gather(acc telegraf.Accumulator) error {
+	timestamp := time.Now().Round(time.Second)
+
+	plugin.retrieveMetrics(acc, timestamp)
+
+	return nil
+}
+
+func (plugin *T128GraphQL) buildQuery() string {
+	//allow multiple inputs like names[ComboEast,ComboWest] ?
+	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
+	query := "query MyQuery{" + replacer.Replace(plugin.EntryPoint) + "{"
+
+	for _, element := range plugin.Fields {
```

Comment:
Oh I see. You used that example because you don't have support for the nested fields/tags yet. That might get a little tricky and I think you'll need it for device state.

Reply:
I wish there was just a graphql query builder from json in Go so you could just build a struct or something rather than a json string. I've sent you this before but I found this for JS/TS https://github.com/atulmy/gql-query-builder

Reply:
Yeah this is going to be my next big focus. That project should have some good stuff so i'll dig into it to start